Add multi-tenancy usage models blog post (#27200)
kubernetes/website: content/en/blog/_posts/2021-03-24-Kubernetes-Multi-Tenancy-Usage-Models.md

## Multi-tenancy usage models

Kubernetes clusters contain control-plane and worker nodes, with multiple software components on each. Different multi-tenancy usage models describe how these resources are shared across users.  A tenant is defined as a group of users whose workloads are isolated from users that do not belong to the tenant. There are two primary multi-tenancy usage models described below.

### Namespaces-as-a-Service

With the *namespaces-as-a-service* model, tenants share a cluster and tenant workloads are restricted to a set of Namespaces assigned to the tenant. The cluster control plane resources like the API server and scheduler, and worker node resources like CPU, memory, etc. are available for use across all tenants. 

To isolate tenant workloads, each namespace must also contain:
* **role bindings:** for controlling access to the namespace
* **network policies:** to prevent network traffic across tenants

> * **resource quotas:** to limit usage and ensure fairness across tenants

Reviewer: - These could be hyperlinks.
- Isolation between tenants typically needs an extra control: SecurityContextDeny or other restriction on Pod admission (so you can't run a privileged Pod and bypass controls); possibly also PriorityClass to protect important cluster-wide services, and EventRateLimit admission with FlowSchemas.

Author: @stim - thanks! Added the links.

Re: pod security and admission controls, we wanted to emphasize that these are essential independently of multi-tenancy models and hence discussed this in the "Multi-tenancy security considerations" section and a added a call-out for potential multi-tenancy risks. Does that need more emphasis?

Author: @sftim - I believe the restructuring helps address this. Let me know if you recommend additional changes.

---
layout: blog

> title: 'Kubernetes Multi-Tenancy Usage Models'

Reviewer: Do you prefer:
```suggestion
title: 'Two Common Multi-Tenancy Models For Kubernetes'
```
?

date: 2021-03-24
slug: kubernetes-multi-tenancy-usage-models
---

**Authors:** Ryan Bezdicek (Medtronic), Jim Bugwadia (Nirmata), Tasha Drew (VMware),  Fei Guo (Alibaba), Adrian Ludwin (Google)

Kubernetes clusters are typically used by several teams in an organization. Secure sharing of Kubernetes resources allows maximizing productivity and saving costs. In other cases, Kubernetes may be used to deliver software and platform services to end users requiring segmentation and isolation of resources across users.

The Kubernetes Multi-Tenancy Working Group is chartered with defining multi-tenancy models for Kubernetes and making it easier to operationalize multi-tenant use cases.  This blog post, from the multi-tenancy working group members, describes two common usage models for multi-tenancy and introduces working group projects that can be used to simplify Kubernetes multi-tenancy management. 

> We will also be presenting on this content and discussing different use cases at our [Kubecon EU 2021 panel session](https://sched.co/iE66).

Reviewer: ```suggestion
We will also be presenting on this content and discussing different use cases at our Kubecon EU 2021 panel session, [Multi-tenancy vs. Multi-cluster: When Should you Use What?](https://sched.co/iE66).
```

## Multi-tenancy security considerations

Cloud native security involves different system layers and lifecycle phases as described in the [Cloud Native Security Whitepaper](https://www.cncf.io/blog/2020/11/18/announcing-the-cloud-native-security-white-paper/) from CNCF SIG Security.  Without proper security measures implemented across all layers and phases, Kubernetes multi-tenancy can be compromised. 

It is important for any new user to Kubernetes to realize that the default installation of a new upstream Kubernetes cluster is not secure, and you are going to need to invest in hardening it in order to avoid security issues.

At a minimum, the following security measures are required:
* Scan container images for vulnerabilities.
* Apply [CIS Benchmarks for Kubernetes](https://www.cisecurity.org/benchmark/kubernetes/)
* Restrict container image sources to trusted registries.

> * Implement granular [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/).

Reviewer: For internal links, use site-relative URLs:
```suggestion
* Implement granular [RBAC](/docs/reference/access-authn-authz/rbac/).
```

## Multi-tenancy usage models

Kubernetes clusters contain control-plane and worker nodes, with multiple software components on each. Different multi-tenancy usage models describe how these resources are shared across users.  A tenant is defined as a group of users whose workloads are isolated from users that do not belong to the tenant. There are two primary multi-tenancy usage models described below.

### Namespaces-as-a-Service

With the *namespaces-as-a-service* model, tenants share a cluster and tenant workloads are restricted to a set of Namespaces assigned to the tenant. The cluster control plane resources like the API server and scheduler, and worker node resources like CPU, memory, etc. are available for use across all tenants. 

To isolate tenant workloads, each namespace must also contain:
* **role bindings:** for controlling access to the namespace
* **network policies:** to prevent network traffic across tenants
* **resource quotas:** to limit usage and ensure fairness across tenants

> With this model, tenants share cluster-wide resources like Cluster Roles and Custom Resource Definitions (CRDs) and hence cannot create or update these cluster-wide resources.

Reviewer: We usually don't put spaces inside API kinds, eg:
```suggestion
With this model, tenants share cluster-wide resources like ClusterRoles and CustomResourceDefinitions (CRDs) and hence cannot create or update these cluster-wide resources.
```

* Apply [CIS Benchmarks for Kubernetes](https://www.cisecurity.org/benchmark/kubernetes/)
* Restrict container image sources to trusted registries.
* Implement granular [RBAC](/docs/reference/access-authn-authz/rbac/).
* Apply default “deny-all” [network policies](/docs/concepts/services-networking/network-policies/).
* Implement [Kubernetes Pod Security Standards](/docs/concepts/security/pod-security-standards/).

Additional recommendations include:
* Use a policy engine to validate configurations 
* Implement runtime scanning
* For stronger isolation, consider VM-based container sandboxing 

> While proper security is required independently of tenancy models, not having essential security controls like [pod security](https://kubernetes.io/docs/concepts/security/pod-security-standards/) in a shared cluster provides attackers with means to compromise tenancy models and possibly access sensitive information across tenants increasing the overall risk profile.

Reviewer 2: ```suggestion
While proper security is required independently of tenancy models, not having essential security controls like [pod security](/docs/concepts/security/pod-security-standards/) in a shared cluster provides attackers with means to compromise tenancy models and possibly access sensitive information across tenants increasing the overall risk profile.
```

Author: @tengqm - thanks! Updated.
